pkg/scheduler/connections/pool.go

```diff
@@ -0,0 +1,157 @@
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	schedulerv1pb "github.com/dapr/dapr/pkg/proto/scheduler/v1"
+	"github.com/dapr/dapr/pkg/scheduler"
+	"github.com/dapr/kit/logger"
+)
+
+var log = logger.NewLogger("dapr.runtime.scheduler")
+
+// Pool represents a connection pool for namespace/appID separation of sidecars to schedulers.
+type Pool struct {
+	Lock             sync.RWMutex
+	NsAppIDPool      map[string]*AppIDPool
+	MaxConnsPerAppID int // future expand to diff conn count for diff appIDs?
+}
+
+// AppIDPool represents a pool of connections for a single appID.
+type AppIDPool struct {
+	lock        sync.RWMutex
+	connections []*Connection
+}
+
+type Connection struct {
+	ConnDetails *scheduler.SidecarConnDetails
+	Stream      schedulerv1pb.Scheduler_WatchJobServer
+}
+
+// Add adds a connection to the pool for a given namespace/appID.
+func (p *Pool) Add(nsAppID string, conn *Connection) {
+	p.Lock.Lock()
+	defer p.Lock.Unlock()
+
+	// Ensure the AppIDPool exists for the given nsAppID
+	if p.NsAppIDPool[nsAppID] == nil {
+		p.NsAppIDPool[nsAppID] = &AppIDPool{
+			lock:        sync.RWMutex{},
+			connections: make([]*Connection, 0),
+		}
+	}
+
+	// Check if adding the connection would exceed the maximum connection count only if it's explicitly set
+	if p.MaxConnsPerAppID != -1 {
+		if len(p.NsAppIDPool[nsAppID].connections) >= p.MaxConnsPerAppID {
+			log.Infof("Sufficient number of Sidecar connections to Scheduler reached. Not adding connection for namespace/appID: %s. Current connection count: %d", nsAppID, len(p.NsAppIDPool[nsAppID].connections))
+			return
+		}
+	}
+
+	// Check if the connection already exists in the pool
+	for _, existingConn := range p.NsAppIDPool[nsAppID].connections {
+		if existingConn.ConnDetails == conn.ConnDetails {
+			log.Infof("Not adding connection for namespace/appID: %s. Connection already exists.", nsAppID)
+			return
+		}
+	}
+
+	p.NsAppIDPool[nsAppID].lock.Lock()
+	defer p.NsAppIDPool[nsAppID].lock.Unlock()
+	p.NsAppIDPool[nsAppID].connections = append(p.NsAppIDPool[nsAppID].connections, &Connection{
+		ConnDetails: conn.ConnDetails,
+		Stream:      conn.Stream,
+	})
+}
+
+// Remove removes a connection from the pool for a given namespace/appID.
+func (p *Pool) Remove(nsAppID string, conn *Connection) {
+	p.Lock.Lock()
+	defer p.Lock.Unlock()
+	if id, ok := p.NsAppIDPool[nsAppID]; ok {
+		id.lock.Lock()
+		defer id.lock.Unlock()
+		for i, c := range id.connections {
+			if c.ConnDetails == conn.ConnDetails {
+				id.connections = append(id.connections[:i], id.connections[i+1:]...)
+				break
+			}
+		}
+	}
+}
+
+// WaitUntilReachingMaxConns waits until the minimum connection count (of sidecars) is reached for a given namespace/appID.
+func (p *Pool) WaitUntilReachingMaxConns(ctx context.Context, nsAppID string, maxConnPerApp int, maxWaitTime time.Duration) error {
+	timeout := time.After(maxWaitTime)
+
+	for {
+		p.Lock.RLock()
+		appIDPool, ok := p.NsAppIDPool[nsAppID]
+		p.Lock.RUnlock()
```

Comment:
move down p.Lock.RUnlock() after ok check and maybe p.Lock.RUnlock() after 102: appIDPool.lock.RUnlock()

Reply:
I rm-ed the extra lock entirely